Sources/Features/Breadcrumb/BacktraceBreadcrumbFileHelper.swift

```diff
@@ -36,78 +32,75 @@ enum BacktraceBreadcrumbFileHelperError: Error {
-        let text: String
-        do {
-            text = try convertBreadcrumbIntoString(breadcrumb)
-        } catch {
-            BacktraceLogger.warning("\(error.localizedDescription) \nWhen converting breadcrumb to string")
-            return false
-        }
-
-        let textBytes = Data(text.utf8)
-        if textBytes.count > maximumIndividualBreadcrumbSize {
-            BacktraceLogger.warning(
-                "Discarding breadcrumb that was larger than the maximum specified (\(maximumIndividualBreadcrumbSize).")
-            return false
-        }
-
         do {
-            try dispatchQueue.sync {
-                // Keep removing until there's enough space to add the new breadcrumb (leaving 512 bytes room)
-                while (queueByteSize() + textBytes.count) > (maxQueueFileSizeBytes - 512) {
-                    try queue.pop(1, error: ())
+            // Serialize breadcrumb: [String: Any] into Data
+            let breadcrumbJsonData = try JSONSerialization.data(withJSONObject: breadcrumb)
+            // Serialize Data into a JSON string
+            guard let breadcrumbJsonString = String(data: breadcrumbJsonData, encoding: .utf8) else {
+                BacktraceLogger.warning("Error when adding breadcrumb to file")
+                return false
+            }
+            // Calculate the size of the breadcrumb and add it to queue
+            let breadcrumbSize = breadcrumbJsonData.count
+            // Check if breadcrumb size is larger than the maximum specified
+            if breadcrumbSize > maximumIndividualBreadcrumbSize {
+                BacktraceLogger.warning(
+                    "Discarding breadcrumb that was larger than the maximum specified (\(maximumIndividualBreadcrumbSize).")
+                return false
+            }
+            // Store breadcrumb Json String and size in Dictionary [String : Any]
+            let queueBreadcrumb = ["breadcrumbJson": breadcrumbJsonString, "size": breadcrumbSize] as [String : Any]
+            // Queue breacrumb
+            queue.enqueue(queueBreadcrumb)
+            // Iterate over the queue from newest to oldest breadcrumb and build an array of encoded strings
+            var breadcrumbsArray = [String]()
+            var size = 0
+            for index in (0..<queue.count).reversed() {
+                guard let queueBreadcrumb = queue.allElements()[index] as? [String: Any] else {
+                    BacktraceLogger.warning("Error weh fetching breacrumbs from queue")
+                    return false
+                }
+                guard let breadcrumbSize = queueBreadcrumb["size"] as? Int else {
+                    BacktraceLogger.warning("Error when adding breadcrumbSize to array")
+                    return false
+                }
+                // Pop last element if size id greater than maxQueueFileSizeBytes
+                if size + breadcrumbSize > maxQueueFileSizeBytes && !queue.isEmpty {
```

Comment:
we should break here - why? there will be a chance we omit one breadcrumb but add a few older one and this algorithm will be fine with it. It would be better if we will have breadcrumbs in the right order and we won't miss any

Reply:
ok , adding a break here.